### PR TITLE
AWQ resolved mappings -- ensure shapes align

### PR DESCRIPTION
**Pull request by brian-dellabetta**
SUMMARY:
Add a robustness check to AWQ to exclude mappings where the layer shapes don't align. This is a known issue, just wanted to do it in a different PR because it occurs in a different location in AutoAWQ, and I wanted to keep the initial AWQ PR to be as close to the AutoAWQ implementation so the git history reflects our changes well.

TEST PLAN:
Added unit test to check default case and edge cases

### CODE REVIEW

**Comment by kylesayrs** [comment]
Shouldn't we be erroring if the user passes a bad mapping, rather than excluding model-specific names?

**Comment by kylesayrs** [comment]
Also, is there a known way to verify a mapping? How do we know that the model doesn't perform a reshaping in between mappings?

File: src/llmcompressor/modifiers/awq/base.py [hunk 1]
```diff
@@ -166,14 +166,29 @@ def _set_resolved_mappings(self, model: Module) -> None:
...
+
+                        # exclude balance layers whose shapes are incompatible
+                        if (
+                            isinstance(smooth_layer, torch.nn.Linear)
+                            and isinstance(balance_layer, torch.nn.Linear)
+                            and smooth_layer.out_features != balance_layer.in_features
+                        ):
+                            logger.info(
+                                f"Excluding {layer_name} -> {balance_name} "
+                                + "due to shape mismatch"
```
[the remark this answers is not in the document]
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: thanks! never sure with these 😅

**Comment by brian-dellabetta** [comment]
> Shouldn't we be erroring if the user passes a bad mapping, rather than excluding model-specific names?

It does, it will fail in the call if for example a user has `"re:.*input_layernorm"` in the smooth layer mapping but it can't be found. this is an edge case where mapping is ok, but some models don't have compatible shapes

**Comment by brian-dellabetta** [comment]
> Also, is there an architecture-agnostic way to verify a mapping? How do we know that the model doesn't perform a reshaping in between mappings?

I talked to @rahul-tuli about an architecture-agnostic way, but I think for now we resolved to scope this to exactly how the AutoAWQ handles it [here](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/models/llama.py#L57). In reality, if the smooth_layer.out_features doesn't match the balance_layer.in_features, it will fail during runtime. Wonkier architectures may hit further errors, but we resolved to handle those as they arise rather than attempting a more all-encompassing solution

**Comment by kylesayrs** [comment]
@brian-dellabetta I'm worried about breaking/polluting other model architectures. If we're excluding names specific to an architecture, shouldn't we be creating a model architecture mappings registry, similar to smoothquant?

**Comment by brian-dellabetta** [comment]
> @brian-dellabetta I'm worried about breaking/polluting other model architectures. If we're excluding names specific to an architecture, shouldn't we be creating a model architecture mappings registry, similar to smoothquant?

@kylesayrs  we do have a mappings registry, see diff just above [here](https://github.com/vllm-project/llm-compressor/pull/1372/files#diff-d9225731e68619b86c07c90cbe39639b92443c7dc7157e6b32571e29fd49c736L32), for each model family. However, the specific model can have different shapes. for example, `["TinyLlama/TinyLlama-1.1B-Chat-v1.0", "meta-llama/Llama-2-7b-hf", "meta-llama/Meta-Llama-3-8B-Instruct"]` can all use the `"Llama"` mappings in the registry, but the v_proj/o_proj need to be excluded for tiny llama and llama 3. Otherwise the user will hit a runtime error. We are using the rule of thumb of trying to adhere to original AutoAWQ implementation for details like this

**Comment by kylesayrs** [comment]
@brian-dellabetta I see. Since `v_proj` and `o_proj` are really common names, I'd still rather limit these changes to the llama model type.

I'd suggest that, rather than logging the exclusion for each layer (which could be noisy), hoisting the logic out of the for loops and modifying the mapping based on the config.

```python3
def validate_mapping(self, model: PreTrainedModel, mappings: List[AWQMapping]):
    if model.config.model_type == "llama":
        num_attn_heads = model.config.num_attention_heads
        num_kv_heads = model.config.num_num_key_value_heads
        if num_attn_heads != num_kv_heads:
            logger.info(f"Excluding v_proj due to shape mismatch")
            mappings = [mapping for mapping in mappings if mapping.smooth_layer != "re:.*v_proj"]
            
     return mappings
```

**Comment by brian-dellabetta** [comment]
> @brian-dellabetta I see. Since `v_proj` and `o_proj` are really common names, I'd still rather limit these changes to the llama model type.

@kylesayrs it is not limited to llama. See AutoAWQ's [qwen](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/models/qwen2vl.py#L48), [mistral](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/models/mistral.py#L58) and [gemma](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/models/gemma.py#L57) implementations for a few additional examples.

**Comment by kylesayrs** [comment]
@brian-dellabetta My point is that AWQ explicitly lists which architectures this rule applies to, whereas the current implementation applies the rule to all architectures.

However, it seems like this is [consistent enough](https://github.com/search?q=repo%3Acasper-hansen%2FAutoAWQ%20%22prev_op%3Dmodule.self_attn.v_proj%2C%22&type=code) across architectures to merit a blanket rule.

If logging for every layer is noisy, we can talk about logging only once.

**Review by kylesayrs — APPROVED** [review]

File: src/llmcompressor/modifiers/awq/base.py [hunk 2]
```diff
@@ -166,13 +166,32 @@ def _set_resolved_mappings(self, model: Module) -> None:
                         balance_name, balance_layer = get_matching_layer(
                             balance_suffix, layer_name, model
                         )
-                        if balance_layer:
-                            balance_layers.append(balance_layer)
-                            balance_names.append(balance_name)
+                        if not (balance_layer):
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:169]: consider removing `()`

**Review by kylesayrs — DISMISSED** [review]

**Review by shanjiaz — APPROVED** [review]
Thanks for adding tests! Logs are much cleaner too. Works for me locally 🎉

**Review by kylesayrs — APPROVED** [review]
